99.md

```diff
@@ -0,0 +1,41 @@
+NIP-99
+======
+
+Nostr Data Apps
+---------------
+
+`draft` `optional` `author:brugeman`
+
+There is a need to have nostr events be generated on-demand by data providers. This NIP defines a best-practice to do that using parameterized replaceable events. The main goal of this NIP is to make sure that on-demand generated events can be queried by clients and stored by relays just like any other nostr events. 
+
+A parameterized replaceable event kind `33333` is defined as a "Nostr Data App" event. Tag `d` is used to specify parameters in a query string format for the event to be generated by the data provider. Data provider's key pair is used to sign the generated events, and the provider's pubkey must be announced and well-known to clients. 
+
+To retrieve an on-demand generated event from a data provider, clients SHOULD send `REQ` message to the provider, containing `kinds:[33333]` filter, `authors:["PROVIDER_PUBKEY"]` filter, and `#d:["params=as&query=string"]` filter.
+
+The data provider SHOULD execute the computations requested by every `#d` tag filter string, and SHOULD return generated nostr events with `content` fields set to the computations' results, with `tags` field set to corresponding params `["d","params=as&query=string"]`, and signed by the provider's key pair.
+
+Clients SHOULD specify the query string in a normalized form to make sure that the same set of parameters produces the same `d` tag. Normalization process: convert parameter names and values into percent-encoded form, sort key value pairs as strings, then build the query string. 
+
+Data provider MAY ignore the requests with non-normalized `d` tags and kind `33333`, and MAY send a proper `NOTICE`.
+
+Data provider MAY send new result events until client terminates the subscription with a `CLOSE` event
```

Comment:
Not sure if its really necessary to mention `CLOSE` and `NOTICE` handling

Reply:
NOTICEs are useful, and custom-relay-writers are lazy (at least I am), so it's a reminder that if you ignore something, at least let the client know. 

The sentence with CLOSE is more about 'send new results to open subs' to remind people that it's not just RPC, it can be more.